include/ygm/io/multi_output.hpp

```diff
@@ -0,0 +1,121 @@
+
+// Copyright 2019-2021 Lawrence Livermore National Security, LLC and other YGM
+// Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <ygm/container/detail/hash_partitioner.hpp>
+
+namespace ygm::io {
+
+namespace fs = std::filesystem;
+
+template <typename Partitioner =
+              ygm::container::detail::hash_partitioner<std::string>>
+class multi_output {
+ public:
+  using self_type = multi_output<Partitioner>;
+
+  multi_output(ygm::comm &comm, const std::string &filename_prefix,
+               bool append = false)
+      : m_comm(comm),
+        pthis(this),
+        m_prefix_path(filename_prefix),
+        m_append_flag(append) {
+    pthis.check(m_comm);
+
+    fs::path prefix_path(filename_prefix);
+
+    // Create directories to hold files
```

Comment:
Should add some checks here to ensure the prefix is a path, does it already exist, etc.